rest/route/stats.go

```diff
@@ -0,0 +1,360 @@
+package route
+
+// This file defines the handlers for the endpoints to query the test and task execution statistics.
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/stats"
+	"github.com/evergreen-ci/evergreen/rest/data"
+	"github.com/evergreen-ci/evergreen/rest/model"
+	"github.com/evergreen-ci/gimlet"
+	"github.com/pkg/errors"
+)
+
+const (
+	// Requester API values
+	requesterMainline = "mainline"
+	requesterPatch    = "patch"
+	requesterTrigger  = "trigger"
+	requesterAdhoc    = "adhoc"
+
+	// Sort API values
+	sortEarliest = "earliest"
+	sortLatest   = "latest"
+
+	// GroupBy API values
+	testGroupByDistro  = "test_task_variant_distro"
+	testGroupByVariant = "test_task_variant"
+	testGroupByTask    = "test_task"
+	testGroupByTest    = "test"
+
+	// API Limits
+	maxGroupNumDays = 26 * 7 // 26 weeks which is the maximum amount of data available
+	maxNumTests     = 50
+	maxNumTasks     = 50
+
+	// Format used to encode dates in the API
+	dateFormat = "2006-01-02"
+)
+
+type testStatsHandler struct {
+	sc     data.Connector
+	filter stats.StatsFilter
+}
+
+func (tsh *testStatsHandler) Factory() gimlet.RouteHandler {
+	return &testStatsHandler{sc: tsh.sc}
+}
+
+func (tsh *testStatsHandler) Parse(ctx context.Context, r *http.Request) error {
+	vals := r.URL.Query()
+	project := gimlet.GetVars(r)["project_id"]
+	if project == "" {
+		return gimlet.ErrorResponse{
+			Message:    "projectId cannot be empty",
+			StatusCode: http.StatusBadRequest,
+		}
+	}
+	tsh.filter = stats.StatsFilter{Project: project}
+
+	return parseTestStatsFilter(vals, &tsh.filter)
+}
+
+func (tsh *testStatsHandler) Run(ctx context.Context) gimlet.Responder {
+	var err error
+	var testStatsResult []stats.TestStats
+
+	testStatsResult, err = tsh.sc.GetTestStats(&tsh.filter)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrap(err, "Failed to retrieve the test stats"))
```

Comment:
Do you mean the error returned by the `tsh.sc.GetTestStats()` method?
If so, the answer is no.

Reply:
maybe want to offer more specific errors from the getTestStats to override the default error status.

Reply:
We don't expect errors from this call because of user input so they will all trigger 500 responses. I made changes to use MakeJSONInternalErrorResponder instead.